Add tests for frontend.

meandavejustice opened the pull request:
Tests will be written in [tape](https://github.com/substack/tape)
They are run with locally with [testling-cli](https://github.com/substack/testling)
They are piped to [faucet](https://github.com/substack/faucet) for pretty output.

todo:
- [x] Get remaining tests to pass (waiting on a change in #371)
- [x] set up coverage stats (coverify maybe)
- [x] add documentation for testing
- [x] add option to run tests in browser(they are currently running headless only)
- [x] hook up to run on travis.

The list of tests are [here](https://github.com/mozilla/idea-town/issues/147#issuecomment-168237808)

cc/ @lmorchard @chuckharmston @6a68 for discussion.

lmorchard commented on idea_town/frontend/static-src/test/general.js:
```diff
@@ -0,0 +1,21 @@
+var path = require('path');
+var fs = require('fs');
+
+test(`🖳  Running Tests for ${__filename} ␄`, function(a) {
```
Hmm, node v0.12 on Travis seems to dislike some characters in this line.

Node on my machine dislikes them too (v0.12.7)

Oh, actually I think it might be an issue with the ES6 template string... 0.12 doesn't support them. Are you running these tests in one of the Docker containers? If not you may have a different version of node than me & travis.

v0.12 isn't exactly a new version of node, so we can look into upgrading in Docker & on Travis... but that might be a task out of scope for this particular issue

meandavejustice replied:
I'm removing them for this test. Since this test is run outside of the babel transform.

lmorchard commented on idea_town/frontend/static-src/app/views/#experiment-page.js#:
```diff
@@ -0,0 +1,249 @@
+import app from 'ampersand-app';
```
Is this file meant to be here? `#experiment-page.js#` seems like maybe a backup filename or something.

meandavejustice replied:
whoops, nice catch. Need to add all of these to my global gitignore.

lmorchard commented on TESTING.md:
```diff
@@ -0,0 +1,24 @@
+The client tests are written with [tape](https://www.npmjs.com/package/tape), and
+run with [testling](https://www.npmjs.com/package/testling).
+
+Here are a couple of tips to helpout when testing clientside code:
```
Nits:
`s/helpout/help out/`
`s/clientside/client-side/`

jaredhirsch commented on CONTRIBUTING.md:
```diff
@@ -70,7 +77,7 @@ Reviewers will run through roughly the following checklist:
 - Are error cases handled?
 
-(Once we have some tests, the reviewer will check that all the tests are passing,
+(the reviewer will check that all the tests are passing,
```
maybe just pull this into the list of bullet points above. speaking of which, does that list of reviewer things to look for actually track with how reviews are done? if not, should we modify the list or just remember to use the list when reviewing? cc @lmorchard

meandavejustice replied:
I guess _travis_ checks to see that the tests are run.

jaredhirsch replied:
> I guess travis checks to see that the tests are run.

For sure, but _humans_ need to verify that the tests make sense, are reasonably independent of one another (so that one bug doesn't cause lots of seemingly-unrelated tests to fail), and that the most important or tricky corners have a little extra thorough coverage.

jaredhirsch commented on gulpfile.js:
```diff
@@ -100,6 +101,8 @@ gulp.task('scripts', ['lint'], function scriptsTask() {
   return bundledStream;
 });
 
+gulp.task('test', shell.task('npm run test:browser:live'));
```
It seems funny to use both gulp and npm for running the same set of tasks. Is there any reason we want it in both places, or can we standardize on one?

meandavejustice replied:
Thanks for bringing this up. I should probably move over to only using gulp.

jaredhirsch commented on idea_town/frontend/static-src/test/lib/router.js:
```diff
@@ -0,0 +1,100 @@
+
+    if (evt.page === 'home') {
+      t.ok(evt.page, 'Should redirect to home if logged in with creds');
+      app.reset();
```
If `app.reset` needs to be called after every test completes, you'll need to make sure it gets called even if the test fails or throws. For this reason, it's a good idea to extract it out to a test runner.

jaredhirsch commented on idea_town/frontend/static-src/test/lib/router.js:
```diff
@@ -0,0 +1,100 @@
+import Me from '../../app/models/me';
+import Experiments from '../../app/collections/experiments';
+
+test(`🖳  Running Tests for ${__filename} ␄`, a => a.end());
```
Seems like this bit of boilerplate could be extracted out to a runner or parent class, too.

jaredhirsch commented on idea_town/frontend/static-src/test/lib/router.js:
```diff
@@ -0,0 +1,100 @@
+      id: 'gary@busey.biz',
+        addon: {url: '/lol nothing matters'}
+      }
+  });
```
Looks like quite a bit of duplicated test setup in this file. Can you extract this to a setup method?

jaredhirsch commented on idea_town/frontend/static-src/test/views/contributor-view.js:
```diff
@@ -0,0 +1,9 @@
+
+test(`🖳  Running Tests for ${__filename} ␄`, a => a.end());
+
+test('Contibutor view renders', t => {
```
'Contributor'

meandavejustice committed:
fixes #147: add tests for the client, running in browser and headless,
add TESTING.md documentation for tests, add reference to TESTING.md to
CONTRIBUTING.md. coverage support with npm script `test:cov`,run tests
in browser with `test:browser`, watch with `test:browser:live`, run
headless with `test:client`, general tests with`test:general` and all
with `test`.

Added Dependencies include:
tape - tap test runner.
tape-dom - tape dom reporter
testling - headless test runner
require-globify - require file globs(used for browser tests)
faucet - tape pretty formatter(reporter)
budo - tiny dev server for browser tests
coverify - coverage reports

Updated tox.ini for running headless tests in travisci

jaredhirsch commented on idea_town/experiments/models.py:
```diff
@@ -42,6 +43,11 @@ class Meta:
+    def installation_count(self):
+        return UserInstallation.objects.distinct('user').filter(
+            experiment=self).count()
+
```
Hmm, is this supposed to be here? Seems like a rebase might be needed?

meandavejustice replied:
woah nice catch. lemme get rid of this

jaredhirsch commented on idea_town/frontend/static-src/app/views/base-view.js:
```diff
@@ -33,7 +33,7 @@ export default AmpersandView.extend({
     // HACK: Slap the same data-l10n-args data on every localized node, because
     // the most common case is they all need the same model data.
-    const nodes = this.el.querySelectorAll('[data-l10n-id]');
+    const nodes = this.queryAll('[data-l10n-id]');
```
It's fine if you want to include this bug fix in this PR, but it needs to be in a separate commit.

There are tons of ways to do this, but here's one way I'd do it:

``` sh
# pull this file out of the main commit
git reset HEAD~1 idea_town/frontend/static-src/app/views/base-view.js
# amend the main commit
git commit --amend --no-edit
# create a new commit
git add idea_town/frontend/static-src/app/views/base-view.js
git commit -m "Use an array, not a NodeList, to allow for-of iteration of l10n nodes"
```

From there, you can `git rebase -i HEAD~2` and swap the order of the two commit lines to put the bug fix earlier in history than the main testing commit.

meandavejustice replied:
Thanks @6a68 I'll take care of that :)

meandavejustice replied:
https://github.com/meandavejustice/idea-town/commit/454e232792a4cf8dbb07bc87e89338ea47be1519

meandavejustice committed:
Use an array, not a NodeList, to allow for-of iteration of l10n nodes
